ouroboros-consensus/docs/diffusion-pipelining-design/block-diffusion-pipelining.tex

```diff
@@ -0,0 +1,554 @@
+use a later one, then the earlier ones can no longer be used again.
+
+\subsection{Many valid block bodies}
+The case for many block bodies is straightforward: they make no difference to
+chain selection. So there being multiple different blocks with different
+content does not alter the informal argument above. It is certainly adversarial
+behaviour for a single block producer to create multiple different blocks in a
+slot, but it does not directly create more work for the honest nodes. A node that
+has adopted a chain, will not select another from the same producer since it is not
+better from a chain selection ordering perspective.
+
+\subsection{Many invalid block bodies}
+\label{many-invalid-block-bodies}
+If we postpone block body validation in the way proposed herein, then all the
+considerations made about "many valid blocks" still apply, and in addition we must
+analyse the new cases for invalid block bodies. For this part of the analysis
+we can presume that the block header is valid.
+
+The bounding argument for the valid block case relies on making strict
+progress in the chain selection ordering, and thus the overall opportunity to make
+progress being bounded. The immediate difficulty in trying the same argument
+with invalid block bodies is that once we discover they are invalid, and we
+stick with our current chain, then we have not made any progress in the chain
+selection ordering. We return to the same state where we are prepared to
+consider fresh (or perhaps even the same) invalid blocks.
+
+Under the full prompt validation model when we encounter an invalid block we
+can promptly cut off all peers that gave us that invalid block (even peers
+that just gave us the header). We cannot do that under the delayed block body
+validation proposal because we can no longer pin the blame on the immediate
+peer for failing to validate the block body before sending it to us.
+
+Fortunately this problem does not appear insurmountable. We can still pin the
+blame for a bad block on the issuer of the block. The block issuer is typically
+not a direct peer therefore we cannot use disconnection as a solution. If however, we simply
+employ some state to temporarily store the block issuer details, then we can use that reference
+to not consider any more blocks signed by this issuer in this slot. They
+have demonstrated themselves to be adversarial by signing a bad block, and we
+shouldn't spend any further resources validating blocks from them in the short term. 
+Note that for this purpose we must identify the block issuer by the hot key certificate and 
+not the `cold' permanent identity. This is due to the fact that a compromised hot key could 
+be used to sign bad blocks, and that ought to still be a situation from which a block producer can 
+recover from by signing a new certificate with a higher issue number.
+
+It appears that it would be sufficient to hold a state of bad block issuers
```

Comment:
> only until the current chain is successfully extended

I _think_ that rule addresses a specific nuance. And I think that nuance should be explained more explicitly.

Suppose A, B, and C are all connected to each other, all on the same chain, the adversary A is sole leader of the next slot, and the adversary A mints and sends an invalid block ab to B and mints and send a valid block ac to C. B will ban A's hot cert and then C will offer to send the (valid) block ac it got from A to B, who will refuse to download it, since the hot cert is the same as that of the ab it already discovered is invalid. This steady state is alarming, since two honest nodes B and C are now in a steady-state but on different chains (specifically, in this example, C is one block (=ac) ahead of B).

The upshot is that B may need to fetch and download a block (eg the block ac) from a hot cert it has banned, but only if a not-yet-banned issuer has extended such block (eg if C were to mint a block that extends ac).

It's not terribly clear in the current text that "until .... successfully extended" in particular means "until I receive a valid header that extends a block I had so far been refusing because of a hot cert ban".